### ISSUE TITLE
Make specified row visible in table

### ISSUE THREAD

**Issue by MonJamp** [ocornut/imgui#3692]
**Version/Branch of Dear ImGui:**

Version: 1.80 WIP (17909)
Branch: master

**My Issue/Question:**

A useful feature for the tables API would be to have a way to ensure that a specific row is visible. An example of the API I'm requesting can be seen in wxWidgets where there is a function called `Focus` and `EnsureVisible`. These functions take in the index of the row and check if the row is currently visible. If the row is not currently visible then the table is scrolled until the row is in view.

Here is a hacky proof of concept, where I've implemented scrolling to a selected row:
```c
struct FocusListClipper : public ImGuiListClipper
{
	void Focus(int row_index)
	{
		ImGuiContext& g = *ImGui::GetCurrentContext();
		ImGuiWindow* window = g.CurrentWindow;
		// Item height is hard coded here
		int item_height = 17;
		window->Scroll.y = IM_FLOOR(item_height * row_index);
	}
};
```
```c
FocusListClipper clipper;
clipper.Begin(disasmData.size());

if (focus_address)
{
	clipper.Focus(address_index);
	focus_address = false;
}

while (clipper.Step())
{
	for (int i = clipper.DisplayStart; i < clipper.DisplayEnd; i++)
	{
		...
```

This is the code I'm using to implement a debugger which can step through disassembled code. The `Focus` function ensures that each time I step to the next line of code, the next line of code will still be visible. I'm using it for a rather niche purpose however I'm sure there are many other uses. For example this could be used to search for an item in a table and scrolling that item into view.

**Comment by ocornut** [comment]
Writing some ideas but note this is going in quite many direction:

Altering `window->Scroll.y` after `Begin()` is undefined but you can call `SetScrollY()` and that new scroll value will be applied on the first `Begin()` of the next frame. Or use `SetNextWindowScroll()` before `Begin()`.

That value of 17 is likely to be `== GetTextLineHeightWithSpacing()`. For framed items it will be `== GetFrameHeightWithSpacing()`. Since setting `window->Scroll.y` is likely to not apply at the end of the frame you may as well just do it after the clipper loop and use `clipper.ItemsHeight`. You might as well realize that your inherited class doesn't use the clipper at all, so you may just replace your code by:
```
(after the clipper loop)
if (focus_address)
   ImGui::SetScrollY(clipper.ItemsHeight * address_index);
focus_address = false;
```
or
```
(before the child Begin)
if (focus_address)
   ImGui::SetNextWindowScroll(ImVec2(-1.0f, ImGui::GetTextLineHeightWithSpacing() * address_index);
focus_address = false;
```

Generally speaking, `SetScrollHereY()` is designed to make a given current position visible, that but enforce centering, while imgui_internal.h's `ScrollToBringRectIntoView()` doesn't.  `SetScrollHereY()` is also not easy to use with a clipper. And `ScrollToBringRectIntoView()` currently doesn't handle frozen rows correctly. Notice how Demo>Scrolling> uses `SetScrollY()` and `SetScrollFromPosY()`. Ref #2900.

TL;DR; I need we need to:
- Consolidate SetScrollHereY + ScrollToBringRectIntoView() into a new more flexible function with flags (as "making item centered" vs "scrolling just enough to make visible" are both are desirable in different situations)
- Fix `ScrollToBringRectIntoView()` to take account of frozen rows. The reason it is broken is those scroll function are currently relying on title bar height and menu bar height as hard-coded decoration sizes for a the non-scrolling part (the "menu layer") and we need to make that more flexible. But this requires a bit of a refactor as currently decoration sizes are locked during Begin() and we want to allow them to be extended later. When we do that along with fixing that function for frozen rows in tables, we can make it easier to setup custom status bars, and we can even remove the `ImGuiWindowFlags_MenuBar` flag completely as `BeginMenuBar()` could request the space dynamically.
- Make it possible to request of the clipper to make a given region always unclipped to allow items `SetScrollHere` types of functions.

Sorry this may be all a little bit cryptic, writing is mainly in order to leave a trail.

**Comment by MonJamp** [comment]
Thank you, that was really helpful. Didn't expect the quick response. I just started using imgui a few days ago and I was trying to study the API to figure out the best way to implement this feature.
